File: others/uber/Parking-Lot/parking_lot.py
MOTORCYCLE = 0
COMPACT = 1
LARGE = 2


class ParkingLot:
    def __init__(self, motorcycle, compact, large):
        # Available counts by spot type index.
        self.spots = [motorcycle, compact, large]
        # Vehicle id -> (spot_type, count) for easy removal.
        self.parked = {}
        # Preferred allocations by vehicle type, in priority order.
        self.rules = {
            "motorcycle": [(MOTORCYCLE, 1), (COMPACT, 1), (LARGE, 1)],
            "car": [(COMPACT, 1), (LARGE, 1)],
            "van": [(LARGE, 1), (COMPACT, 3)],
            "bus": [(LARGE, 1), (COMPACT, 5)],
        }

    def park(self, vid, vtype):
        # Allocate the first rule that fits the remaining capacity.
        if not vid or vid in self.parked or vtype not in self.rules:
            return False
        for spot, count in self.rules[vtype]:
            if self.spots[spot] >= count:
                self.spots[spot] -= count
                self.parked[vid] = (spot, count)
                return True
        return False

    def remove(self, vid):
        # Free the previously allocated spots for this vehicle id.
        alloc = self.parked.pop(vid, None)
        if alloc is None:
            return False
        spot, count = alloc
        self.spots[spot] += count
        return True

    def available(self, spot_type):
        # Remaining count for a given spot type.
        if 0 <= spot_type <= LARGE:
            return self.spots[spot_type]
        return 0


if __name__ == "__main__":
    lot = ParkingLot(1, 4, 1)
    print(lot.park("m1", "motorcycle"))
    print(lot.park("c1", "car"))
    print(lot.park("v1", "van"))
    print(lot.park("b1", "bus"))
    print(lot.remove("v1"))
    print(lot.park("b1", "bus"))

File: others/uber/Parking-Lot/ParkingLot.go
package uber

// spotType
const (
	SpotMotorcycle = iota
	SpotCompact
	SpotLarge
)

// vehicleType
const (
	Motorcycle = "motorcycle"
	Car        = "car"
	Van        = "van"
	Bus        = "bus"
)

type ParkingLot struct {
	// spots2cnt holds available counts by spot type index.
	spots2cnt [3]int
	// parked maps vehicle id -> (spotType, count) for easy removal.
	parked map[string][2]int
	// rules defines preferred allocations in priority order per vehicle type.
	rules map[string][][2]int
}

// NewParkingLot builds a lot with fixed spot counts and default allocation rules.
func NewParkingLot(motorcycle, compact, large int) *ParkingLot {
	return &ParkingLot{
		spots2cnt: [3]int{motorcycle, compact, large},
		parked:    map[string][2]int{},
		rules: map[string][][2]int{
			Motorcycle: {{SpotMotorcycle, 1}, {SpotCompact, 1}, {SpotLarge, 1}},
			Car:        {{SpotCompact, 1}, {SpotLarge, 1}},
			Van:        {{SpotLarge, 1}, {SpotCompact, 3}},
			Bus:        {{SpotLarge, 1}, {SpotCompact, 5}},
		},
	}
}

// Park tries to allocate a spot (or compact count) based on vehicle rules.
func (p *ParkingLot) Park(id, vehicleType string) bool {
	if _, exists := p.parked[id]; exists {
		return false
	}
	options := p.rules[vehicleType]
	for _, opt := range options {
		spot, count := opt[0], opt[1]
		if p.spots2cnt[spot] >= count {
			p.spots2cnt[spot] -= count
			p.parked[id] = opt
			return true
		}
	}
	return false
}

// Remove frees the spots previously used by the vehicle id.
func (p *ParkingLot) Remove(id string) bool {
	alloc, ok := p.parked[id] // (spotType, count)
	if !ok {
		return false
	}
	p.spots2cnt[alloc[0]] += alloc[1]
	delete(p.parked, id)
	return true
}

// Available returns remaining count for a given spot type.
func (p *ParkingLot) Available(spotType int) int {
	if spotType < 0 || spotType > SpotLarge {
		return 0
	}
	return p.spots2cnt[spotType]
}
